doc/design/context.md

```diff
@@ -0,0 +1,171 @@
+## Context Design
+
+`Net` is the container and controller of a set of `Operator`. Each `Operator` in `Net` has a method called `Run` to make computation. The `Run` method of `Operator` is defined as follows:
+
+```c++
+Error Operator::Run(OpContext* context);
+```
+
+And the `OpContext` is defined as follows:
+
+```c++
+Struct OpContext {
+  Scope* scope;
+  DeviceContext* device_context;
+};
+```
+
+The `Run` method will take `Variable` (containing `Tensor`) from `Scope` to make computation on certain `DeviceContext`. `DeviceContext` provides necessary runtime resources for computation, including CudaStream, cublasHandle and so on.
```

Comment:
I'd suggest that let's target at SingleThreadEngine so to avoid too complex a design that we as a team cannot handle right now. Let's assume that a net runs on a single GPU/CPU by one thread only.  We can keep this design for later upgrading to more complicated engines.